### ISSUE TITLE
The template reference 'myAppInsights' is not valid

### ISSUE THREAD

**Issue by VitorRigoni** [CompositionalIT/farmer#305]
Hello! 👋🏻

I'm going through some trouble trying to deploy an Azure Function and was hoping you'd have any insights.
My issue is actually related to the `link_to_service_plan` and `link_to_app_insights` settings. Turns out I'm trying to link to a service plan and an app insights instance that are located in different resource groups, and I keep getting this error:
`Deployment template validation failed: 'The template reference 'myAppInsights' is not valid: could not find template resource or resource copy with this name.`

I did some digging and, apparently, I have to reference resources located in different resource groups using the resourceId arm function so it takes a resource group name and is properly able to resolve the reference.
I couldn't find a way to make this work from Farmer, though =(

Any ideas how I could make this work?

Thanks!

**Comment by isaacabraham** [comment]
Hey. We've just done something similar for this with app service - people trying to create websites that piggy back on app services in other resource groups. The initial idea of Farmer was for it to own all resources - this kept things simple - but now we have the support inside Farmer to add these "really" external resources it shouldn't be too hard to add support for AI.

Longer term we should look at allowing people to add external resources for any linked item.

**Comment by isaacabraham** [comment]
I don't actually think that you need the external resource group name, just the resource (although I might be wrong here).

**Comment by isaacabraham** [comment]
Linked to #285

**Comment by VitorRigoni** [comment]
Fantastic!

I got some of your code on the linked issue and I got it working for the app service plan! For app insights I worked around by manually setting my app insights instance guid as a setting for the function - this works for us since we have a single static instance of AI for each env.

This is what I got:

```fsharp
let serverFarm = ArmExpression.resourceId(Arm.Web.serverFarms, Farmer.ResourceName "my-server-farm", "my-resourcegroup").Eval()

let myFunction = functions {
  name functionName
  depends_on myDatabase
  depends_on functionStorageAccount
  link_to_storage_account functionStorageAccountName
  link_to_service_plan (Farmer.ResourceName serverFarm)
  app_insights_off
  settings [
    ("connectionString", dbConnectionString)
    ("APPINSIGHTS_INSTRUMENTATIONKEY", "my-app-insights-guid")
  ]
}

let functionResource : Builder =
    fun location -> [
        // Manually "run" the builder to get all the raw ARM resources.
        let resources = (boldWebhooksHandlerFunction :> IBuilder).BuildResources location
        // We know that the first resource is the site
        let rawWebsite = resources.[0] :?> Arm.Web.Site
        // Remove all dependencies from it and return back out
        { rawWebsite with Dependencies = [] }
    ]
```

Thank you so much! 😁

**Comment by isaacabraham** [comment]
Good news - I totally forgot but this is actually already fixed in master and will be released this evening, so you will be able to simply do this in future:

```fsharp
let myWebApp = webApp {
    name "mysuperwebapp"
    link_to_unmanaged_service_plan "my_external_farm" // connect to existing farm
    link_to_unmanaged_app_insights "my_external_ai" // connect to existing ai
}
```

I think that this will give you the flexibility that you're looking for. Going forward, I think we'll look to expose (in an easier way) the internal abstraction that we use to understand how resources refer to one another (trimmed down from `Types.fs`)

```fsharp
/// A ResourceRef represents a linked resource; typically this will be for two resources that have a relationship
/// such as AppInsights on WebApp. WebApps can automatically create and configure an AI instance for the webapp,
/// or configure the web app to an existing AI instance, or do nothing.
type AutoCreationKind<'T> =
    | Named of ResourceName
    | Derived of ('T -> ResourceName)
type ExternalKind =
    | Managed of ResourceName
    | Unmanaged of ResourceName

type ResourceRef<'T> =
    | AutoCreate of AutoCreationKind<'T>
    | External of ExternalKind
```

In other words, a reference to a resource can be *either* auto created by farmer, or external.

* If it's auto-created, it'll either have an explicit name (supplied by the user) or be derived based on the builder e.g. app insights is derived by default based on the web app name.
* If it's external, it's either a managed (farmer-owned) resource that will be deployed within the ARM template, or completely unmanaged and is typically an already-existing resource (which I believe is your case).

**Comment by VitorRigoni** [comment]
That's perfect!

I actually went through the discussion and saw the references to `link_to_unmanaged_service_plan` already merged to master, but there was no new release yet so I was just anxiously waiting for the new version to be able to use that 😁
And meanwhile I can work around with that other way that works fine for me (even though it's pushing the farmer api limits a little bit 😅)

**Comment by isaacabraham** [comment]
Now pushed to NuGet :-)
